src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,208 @@
+
+
+#include <emscripten/emscripten.h>
+#include <emscripten/val.h>
+#include <emscripten/bind.h>
+
+#include <string>
+#include <iostream>
+
+#include "ffi/ir_stream/decoding_methods.hpp"
+#include "spdlog_with_specializations.hpp"
+#include "TimestampPattern.hpp"
+#include "streaming_compression/zstd/Decompressor.hpp"
+#include "ir/LogEventDeserializer.hpp"
+#include "ffi/encoding_methods.hpp"
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+#else
+#define EXTERN
+#endif
+
+// FIXME: use macro
+enum class LOG_VERBOSITY
+{
+    NONE = 0,
+    TRACE,
+    DEBUG,
+    INFO,
+    WARN,
+    ERROR,
+    FATAL
+};
+
+const char *VERBOSITIES[] = {
+    nullptr,
+    "TRACE",
+    "DEBUG",
+    "INFO",
+    "WARN",
+    "ERROR",
+    "FATAL",
+};
+
+EM_JS(void, insert_row, (emscripten::EM_VAL results_handle, const char *message, int64_t timestamp, int verbosity), {
+    const results = Emval.toValue(results_handle);
+    results.push([ UTF8ToString(message), timestamp, verbosity ]);
+});
+
+class ClpIrV1Decoder
+{
+private:
+    clp::streaming_compression::zstd::Decompressor m_zstd_decompressor;
+    std::vector<clp::ir::LogEvent<clp::ir::four_byte_encoded_variable_t>> m_log_events;
+    const int m_timezone_offset_ms{0 * 3600 * 1000};
+
+    std::unique_ptr<clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>> m_deserializer;
+    clp::TimestampPattern m_ts_pattern;
+
+public:
+    ClpIrV1Decoder(const uintptr_t dataBuffer, const int length)
+    {
+        std::cout << "ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length=" << length << std::endl;
+        m_zstd_decompressor.open(reinterpret_cast<const char *>(dataBuffer), length);
+    }
+
+    int buildIdx()
```

Comment:
Do u think it makes more sense to return an error, and make `num_decoded` as a output parameter? In the current design, the error is swallowed by only print the error. There's no way for the upper level caller to be aware of the error

Reply:
According to the [current design in the Log Viewer](https://github.com/y-scope/yscope-log-viewer/blob/0d302c5ef66ea2e0d66ba1981766614f27b04225/new-log-viewer/src/typings/decoders.ts#L50), we will be returning a JS object which includes     `numValidEvents` and `numInvalidEvents`. The "std::errc::result_out_of_range" can be "safely" ignore for now (, or we can create a dummy log event to indicate the stream doesn't have an EOF. What do you think?). For the "std::errc::no_message_available" case, we can throw an exception once we do enable exception details with the compiler, preferably in another PR if there's no objection.